snooty/util_test.py

```diff
@@ -0,0 +1,91 @@
+import sys
+import textwrap
+import xml.etree.ElementTree as ET
+from xml.sax.saxutils import escape
+from typing import Any
+
+__all__ = ("eprint", "ast_to_testing_string", "assert_etree_equals")
+
+
+def eprint(*args: str) -> None:
+    """print() to stderr."""
+    print(*args, file=sys.stderr)
+
+
+class FinalAssertionError(AssertionError):
+    """An AssertionError whose details were already logged to stderr."""
+
+    pass
+
+
+def ast_to_testing_string(ast: Any) -> str:
+    """Create an XML string representation of an AST node."""
+    value = ast.get("value", "")
+    children = ast.get("children", [])
+    attr_pairs = [
+        (k, v)
+        for k, v in ast.items()
+        if k not in ("argument", "value", "children", "type", "position", "options")
+        and v
+    ]
+    attr_pairs.extend((k, v) for k, v in ast.get("options", {}).items())
+    attrs = " ".join('{}="{}"'.format(k, escape(str(v))) for k, v in attr_pairs)
+    contents = (
+        value
+        if value
+        else (
+            "".join(ast_to_testing_string(child) for child in children)
+            if children
+            else ""
+        )
+    )
+    if "argument" in ast:
+        contents = (
+            "".join(ast_to_testing_string(part) for part in ast["argument"]) + contents
+        )
+    return "<{}{}>{}</{}>".format(
+        ast["type"], " " + attrs if attrs else "", contents, ast["type"]
+    )
+
+
+def assert_etree_equals(e1: ET.Element, goal: ET.Element) -> None:
+    """Assert that two XML Elements are the same. If there is a difference in a child,
+       log the difference to stderr."""
+    assert e1.tag == goal.tag
+    if e1.text and goal.text:
+        assert (e1.text.strip() if e1.text else "") == (
+            goal.text.strip() if goal.text else ""
+        )
+    if e1.tail and goal.tail:
```

Comment:
Do we need to take into account if one has a tail and another doesn't?

Reply:
Actually, yes, good point!

Reply:
Updated!